sdks/java/harness/src/main/java/org/apache/beam/fn/harness/PrecombineGroupingTable.java

```diff
@@ -0,0 +1,568 @@
+        return key;
+      }
+
+      @Override
+      public AccumT getValue() {
+        return accumulator;
+      }
+
+      @Override
+      public long getSize() {
+        return keySize + accumulatorSize;
+      }
+
+      @Override
+      public void compact() throws Exception {
+        AccumT newAccumulator = combiner.compact(key, accumulator);
+        if (newAccumulator != accumulator) {
+          accumulator = newAccumulator;
+          accumulatorSize = accumulatorSizer.estimateSize(newAccumulator);
+        }
+      }
+
+      @Override
+      public void add(InputT value) throws Exception {
+        accumulator = combiner.add(key, accumulator, value);
+        accumulatorSize = accumulatorSizer.estimateSize(accumulator);
+      }
+    };
+  }
+
+  /** Adds a pair to this table, possibly flushing some entries to output if the table is full. */
+  @SuppressWarnings("unchecked")
+  @Override
+  public void put(Object pair, Receiver receiver) throws Exception {
+    put(
+        (K) pairInfo.getKeyFromInputPair(pair),
+        (InputT) pairInfo.getValueFromInputPair(pair),
+        receiver);
+  }
+
+  /**
+   * Adds the key and value to this table, possibly flushing some entries to output if the table is
+   * full.
+   */
+  public void put(K key, InputT value, Receiver receiver) throws Exception {
+    Object groupingKey = groupingKeyCreator.createGroupingKey(key);
+    GroupingTableEntry<K, InputT, AccumT> entry = table.get(groupingKey);
+    if (entry == null) {
+      entry = createTableEntry(key);
+      table.put(groupingKey, entry);
+      size += PER_KEY_OVERHEAD;
+    } else {
+      size -= entry.getSize();
+    }
+    entry.add(value);
+    size += entry.getSize();
+
+    if (size >= maxSize) {
+      long targetSize = (long) (TARGET_LOAD * maxSize);
+      Iterator<GroupingTableEntry<K, InputT, AccumT>> entries = table.values().iterator();
+      while (size >= targetSize) {
+        if (!entries.hasNext()) {
+          // Should never happen, but sizes may be estimates...
+          size = 0;
+          break;
+        }
+        GroupingTableEntry<K, InputT, AccumT> toFlush = entries.next();
+        entries.remove();
+        size -= toFlush.getSize() + PER_KEY_OVERHEAD;
+        output(toFlush, receiver);
+      }
+    }
+  }
+
+  /**
+   * Output the given entry. Does not actually remove it from the table or update this table's size.
+   */
+  private void output(GroupingTableEntry<K, InputT, AccumT> entry, Receiver receiver)
+      throws Exception {
+    entry.compact();
+    receiver.process(pairInfo.makeOutputPair(entry.getKey(), entry.getValue()));
+  }
+
+  /** Flushes all entries in this table to output. */
+  @Override
+  public void flush(Receiver output) throws Exception {
+    for (GroupingTableEntry<K, InputT, AccumT> entry : table.values()) {
+      output(entry, output);
+    }
+    table.clear();
+    size = 0;
+  }
+
+  @VisibleForTesting
+  public void setMaxSize(long maxSize) {
```

Comment:
It would be much better if this was a final value and configurable via construction only.